{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# COGS 108 - Final Project (Bioinformatics-Analysis-of-PTBP1-in-Knockout-Mice)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "LINK TO VIDEO: https://drive.google.com/drive/folders/1ycTaOXzr-HL6TftZucL3j9eXL2lDzMxL?usp=sharing\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this project, we analyzed previous datasets on the effect of knocking out PTBP1 in mice to convert astrocytes to neurons using tools in bioinformatics and machine learning. Previous studies have shown evidences that both support and reject the effect of knocking out PTBP1, so we want to reanalyze datasets available to draw new conclusions. Our results indicate that knocking out PTBP1 will create a new cell type that resemebles neuron by looking at genes most differentially expressed in the cells. In conclusion, we think current datasets support the effect of PTBP1 in converting astrocytes to neurons. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Names\n",
    "\n",
    "- Carlos Garcia\n",
    "- Sai Hosuru\n",
    "- Qilin Zhao\n",
    "- Wilson Lubeck\n",
    "- Sharon Zhao"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='research_question'></a>\n",
    "# Research Question"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We are interested in if knocking out the gene PTBP1 will result in conversion from astrocytes to neurons in mouse cortex. This is potentially important for treating neurodenegerative diseases such as Parkinson's Disease and Alzhiemer's Disease since these conditions are due to loss of neurons. We want to use bioinformatics tools to analyze sequencing data to show if removing the gene PTBP1 to increase number of neurons. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id='background'></a>\n",
    "\n",
    "## Background & Prior Work"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Neurodegenerative diseases are dysfunctions caused by progressive loss of structure and function of neurons in the nervous system. In the United States alone, about 50 million patients suffer from such conditions in each year. One common neurodegenerative disease is Parkinson’s Disease (PD) which causes deficits in motor function, depression, and cognitive decline. With PD, patients usually die within 5 to 12 years. Despite the severeness of the condition, the underlying cause of the disease is not well understood, and we lack effective treatment to reverse neurodegeneration. This situation, however, has been potentially changed recently with studies on gene PTBP1 (Zhou, 2020; Qian, 2020). \n",
    "\n",
    "In the nervous system, there are two groups of cells: neuron, which is essential for computation and the foundation of cognition; glial cell, which supports neuronal function by removing metabolic wastes, provide nutrition, spend up signal transduction, etc. A potential path of cure for curing neurodegenerative disease is to regenerate lost neurons by converting glial cells into neurons. Recent studies show that by knockout (removing) PTBP1 can convert astrocytes, a type of glial cell, into functional neurons in both the central nervous system (brain) as well as peripheral nervous system (retina). The regenerated neurons from PTBP1 knockout are also shown to reverse PD symptoms in mouse model. (Qian, 2020) However, there are also controverses over the validity of the research, citing the results were generated by comparing with wrong reference groups as well as low sample number. (Wang, 2021) Therefore, the role of PTBP1 is still inconclusive in mouse models as well as actual human PD patients.\n",
    "\n",
    "An important piece of data used to prove/disprove role of gene PTBP1 in neuronal regeneration is the RNA sequencing dataset. As a result, we propose to re-analyze dataset from multiple papers about PTBP1 to see if neurons are regenerated by removing PTBP1. By using a consistent analysis pipeline, we hope that we could remove variation caused by different analysis approaches.\n",
    "\n",
    "Zhou H, Su J, Hu X, Zhou C, Li H, Chen Z, Xiao Q, Wang B, Wu W, Sun Y, Zhou Y, Tang C, Liu F, Wang L, Feng C, Liu M, Li S, Zhang Y, Xu H, Yao H, Shi L, Yang H. Glia-to-Neuron Conversion by CRISPR-CasRx Alleviates Symptoms of Neurological Disease in Mice. Cell. 2020 Apr 30;181(3):590-603.e16. doi: 10.1016/j.cell.2020.03.024. Epub 2020 Apr 8. PMID: 32272060.\n",
    "\n",
    "Wang LL, Serrano C, Zhong X, Ma S, Zou Y, Zhang CL. Revisiting astrocyte to neuron conversion with lineage tracing in vivo. Cell. 2021 Oct 14;184(21):5465-5481.e16. doi: 10.1016/j.cell.2021.09.005. Epub 2021 Sep 27. PMID: 34582787; PMCID: PMC8526404.\n",
    "\n",
    "Qian, H., Kang, X., Hu, J. et al. Reversing a model of Parkinson’s disease with in situ converted nigral neurons. Nature 582, 550–556 (2020). https://doi.org/10.1038/s41586-020-2388-4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Hypothesis\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We hypothesize knockout PTBP1 does play a role in converting astrocytes into neurons in mice. In the single cell RNA sequencing data, the knockout group will be different from the control group with additional cells resembles neurons."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Dataset(s)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dataset 1\n",
    "- Dataset Name: Blackshaw_Cortex_WT.h5 \n",
    "- Link to the dataset: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM5600669\n",
    "- Number of observations: There are 2847515 cells in the sample.\n",
    "- Description: The dataset is from the paper Hoang. et al (2022). In the paper, they collected cells from the cortex of a wild type mouse and performed single cell RNA sequencing to measure gene expression. The dataset provided above is a h5 file where they show which genes are expressed in which cells as a matrix.\n",
    "\n",
    "Dataset 2\n",
    "- Dataset Name: Blackshaw_Cortex_KO.h5\n",
    "- Link to the dataset: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM5600671\n",
    "- Number of observations: There are 2785834 cells in the sample.\n",
    "- Description: The dataset is from the paper Hoang. et al (2022). In the paper, they collected cells from the cortex of a mouse whose PTBP1 gene is removed. Using single cell RNA sequencing, they measured gene expression. The dataset provided above is a h5 file where they show which genes are expressed in which cells as a matrix.\n",
    "\n",
    "Dataset 3\n",
    "\n",
    "- Dataset Name: Cortical.WT.rep1.txt.gz\n",
    "- Link to the dataset: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE142250\n",
    "- Number of observations: \n",
    "- Description: The file is a txt.gz file that shows the alignment of genes in the sample. It is a compressed file with 12 files. Among the 12 files, we use the 6 files from the cortex. Since this dataset is already aligned and preprocessed, we don’t need to clean the dataset until we compare results generated from the two previous datasets to this as a reference. \n",
    "\n",
    "If you plan to use multiple datasets, add 1-2 sentences about how you plan to combine these datasets.\n",
    "\n",
    "These datasets are individual, therefore there is no need to combine them. We simply compare them in order to do our analysis.\n",
    "\n",
    "Since the datasets are big, we also downloaded them into a google folder for easier access and analysis: https://drive.google.com/file/d/1JSPxJEWoFZstXwWNGi5MAF8a_beBRrjh/view?usp=sharing"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here is the link to the working google colab notebook. In order to run any code here, please check [HERE](https://colab.research.google.com/drive/1vYfouAfEYjac_zMl24wUU5HxO5cIQsj7?usp=sharing)\n",
    "\n",
    "We simply import all the libraries we will work with, as well as a bash script to install miniconda as it is used for many bioinformatics pipelines.\n",
    "\n",
    " *NOTE* :\n",
    "The COGS108 jupyterhub doesn't allow us to install the conda package or scanpy, so much of this code is ran on google colab or personal jupyter notebooks.\n",
    "All the code is presented here, but the google colab notebook that *ACTUALLY* runs is linked at the top. If anything fails to run here, check the colab link attached above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "MINICONDA_INSTALLER_SCRIPT=Miniconda3-4.5.4-Linux-x86_64.sh\n",
    "MINICONDA_PREFIX=/usr/local\n",
    "wget https://repo.continuum.io/miniconda/$MINICONDA_INSTALLER_SCRIPT\n",
    "chmod +x $MINICONDA_INSTALLER_SCRIPT\n",
    "./$MINICONDA_INSTALLER_SCRIPT -b -f -p $MINICONDA_PREFIX\n",
    "\n",
    "\n",
    "%%bash\n",
    "pip install scanpy\n",
    "pip install matplotlib"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import gzip\n",
    "import os\n",
    "import anndata as ad"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Cleaning"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " 1. Use scanpy to filter out any cells that expresses less than 200 genes\n",
    " 2. Using scanpy, Remove genes that are expressed in less than 5 cells. \n",
    " 3. Change index column into a sequence column, in order to allow for easier access.\n",
    " 4. Label genes in dataframe that starts with MT, which denotes genes that are expressed in mitochondria\n",
    " 5. Quantify and plot processed data to ensure proper handling.\n",
    "\n",
    " In order to make the changes easier, we ran a google colab notebook as there was issues "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "blackshaw_ko = sc.read_10x_h5('./dataset/blackshaw_data/Blackshaw_Cortex_KO.h5')\n",
    "blackshaw_wt = sc.read_10x_h5('./dataset/blackshaw_data/Blackshaw_Cortex_WT.h5')\n",
    "\n",
    "sc.pp.filter_cells(blackshaw_ko, min_genes = 200)\n",
    "sc.pp.filter_genes(blackshaw_ko, min_cells = 5)\n",
    "\n",
    "sc.pp.filter_cells(blackshaw_wt, min_genes = 200)\n",
    "sc.pp.filter_genes(blackshaw_wt, min_cells = 5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "blackshaw_ko.obs.reset_index(inplace=True)\n",
    "blackshaw_ko.obs = blackshaw_ko.obs.rename(columns={'index':'sequence'})\n",
    "\n",
    "#we want to remove mitochondrial genes since it is not effective in determining what type of cell\n",
    "#we are looking at\n",
    "blackshaw_ko.var['mt'] = blackshaw_ko.var_names.str.startswith('MT-')\n",
    "blackshaw_ko.var_names_make_unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.calculate_qc_metrics(blackshaw_ko, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "\n",
    "#visualize analyzed data to make sure the cleaning is effective.\n",
    "sc.pl.violin(blackshaw_ko, ['n_genes_by_counts', 'total_counts', 'pct_counts_mt'], jitter=0.4, multi_panel=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Results:\n",
    "\n",
    "The result for data cleaning is quantification.png in this link: https://drive.google.com/drive/folders/1AVPFA20mcXtbwaeB7AK89gA-cUtS2t7e?usp=sharing\n",
    "\n",
    "The plots (quantification.png; the plots from left to right are labeled as 1 to 3) show the distribution of number of genes expressed in each cell, the number of cells each gene is expressed in, and the percentage of mitocondrial genes in each cells. For the first plot, we see that there are no cells that express less than 200 genes; for the second plot, there is no gene expressed in less than 5 cells; for the last plot, we see all cells have 0% mitocondrial genes. These results indicate our cleaning is effective. One concern we discovered, however, is the uneven distribution of cell and genes (plot 1 and 2). The dsitribution is strongly skewed toward less genes express and less cells a gene is expressed in. This is concerning to us because it could indicate the sensitivity of sequencing is low, but with a fairly large number of dataset (2847515 and 2785834 cells), we should still get descent number of reads for highly expressed gene and cells with high expression level. We also don't want to be overly selective to eliminate genes and cells that might be indicative of certain cell types. Therefore, we decided to keep the parameters for lowest gene copy and cell number as 200 and 5."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Data Analysis & Results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For our EDA, using the dataset we had previously cleaned, we wanted to know how many cells we were working with, and which genes were differentially expressed. Again, everything was run on a google colab notebook, which is linked here: <link>\n",
    "\n",
    "For the analysis, we first normalized the expression level across cells, so the same gene expressed in differerent cells can be compared. Then, to avoid extreme values for genes that are extremely abundant, we take the log of the expression."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## YOUR CODE HERE\n",
    "## FEEL FREE TO ADD MULTIPLE CELLS PER SECTION\n",
    "sc.pp.normalize_total(blackshaw_ko, target_sum = 1e4)\n",
    "sc.pp.log1p(blackshaw_ko)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, with the built in function high_variable_genes, we select the most variable genes across groups which will be used later for clustering. Then, we calculate the principle components and the variance each represent. The result is shown in graph PCA. Here, we can see that the variance dropped significantly after PC7 and PC20, so we tried to cluster with the first 7 and the first 20 principle components. The results are shown in images 7_PCs and 20_PCs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#we find the most highly variable genes across different clusters. this allows us\n",
    "#to find markers that are most differentially expressed across cell types .\n",
    "sc.pp.highly_variable_genes(blackshaw_ko)\n",
    "reads_ko = blackshaw_ko[:, blackshaw_ko.var.highly_variable]\n",
    "sc.tl.pca(reads_ko, svd_solver='arpack')\n",
    "sc.pl.pca_variance_ratio(reads_ko, log=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#here, we take only the first 7 principle components to cluster cells because the variance of\n",
    "sc.pp.neighbors(reads_ko, n_neighbors=15, n_pcs=7)\n",
    "sc.tl.umap(reads_ko)\n",
    "sc.pl.umap(reads_ko)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#clustering based on 20 principle components.\n",
    "sc.pp.neighbors(reads_ko, n_neighbors=15, n_pcs=20)\n",
    "sc.tl.umap(reads_ko)\n",
    "sc.pl.umap(reads_ko)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The results show that clustering improves significantly with 20 PCs than 7 PCs. With 20 PCs, the clusters are more clearly seperated, and there are more disjoint clusters in the graph, so the resolution is also higher."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.leiden(reads_ko)\n",
    "sc.pl.umap(reads_ko, color=['leiden'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We also plotted the top 10 most differentially expressed genes in each clusters. These genes will be analyzed later for cell type identification."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#find highly variable genes among each cluster.\n",
    "sc.tl.rank_genes_groups(reads_ko, groupby='leiden', n_genes=10, method = 'wilcoxon')\n",
    "sc.pl.rank_genes_groups(reads_ko, n_genes = 10, sharey=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "With the PTBP1 knockout sequencing data analyzed, we also looked into the wild type animals. The analysis performed is the same as the knockout aniaml. We also extracted the top 10 most differentially expressed genes in each clusters in the wild type data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#same analysis is performed for wild type animal\n",
    "sc.pp.normalize_total(blackshaw_wt, target_sum = 1e4)\n",
    "sc.pp.log1p(blackshaw_wt)\n",
    "sc.pp.highly_variable_genes(blackshaw_wt)\n",
    "reads_wt = blackshaw_wt[:, blackshaw_wt.var.highly_variable]\n",
    "sc.tl.pca(reads_wt, svd_solver='arpack')\n",
    "sc.pl.pca_variance_ratio(reads_wt, log=True)\n",
    "\n",
    "sc.pp.neighbors(reads_wt, n_neighbors=15, n_pcs=20)\n",
    "sc.tl.umap(reads_wt)\n",
    "sc.pl.umap(reads_wt)\n",
    "\n",
    "sc.tl.leiden(reads_wt)\n",
    "sc.pl.umap(reads_wt, color=['leiden'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.tl.rank_genes_groups(reads_wt, groupby='leiden', n_genes=10, method = 'wilcoxon')\n",
    "sc.pl.rank_genes_groups(reads_wt, n_genes = 10, sharey=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Ethics & Privacy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since we are using the data collected from mice, we have considered the ethics of using mice in scientific experiments. These datasets from the papers are publicly available on the NCBI website. Since the authors declare no financial interests in these datasets, we do not need to worry about conflict of interests. \n",
    "\n",
    "Another concern of ethics is the animal protocols they used for the experiments. Since the protocols are reviewed and verified by the institutions these research groups are at (UCSD and Johns Hopklins University) following federal guidelines, we believe the animals are treated ethically during experiments."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Conclusion & Discussion"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Our analysis agrees with the conclusion that knocking out PTBP1 generates neurons in mice cortex. With the same analysis parameters, there are 30 cell clusters in the knockout dataset (image: knockout_clusters.png) compared to 27 clusters in the wild type dataset (image: wildtype_clusters.png). We then compared the clusters from the two groups to see the characteristics of the three extra clusters in the wildtype dataset. These clusters are cluster 24, 26, and 28 in image knockout_clusters.png. Among these clusters, cluster 28 differentially express the gene Notch3 compared to other clusters. Notch3 is a marker gene for neural progenitor cells which are precursors for neurons in the brain. Since this cluster is not present in the wild type dataset, we conclude cluster 28 indicates neurons formed by knockout PTBP1. This furthur proves our hypothesis that by knockout PTBP1, we can regenerate neurons from astrocytes.\n",
    "\n",
    "Our conclusion agree with some previous works but not others. There could be various reasons for this difference in conclusion. Firstly, our analysis protocol could use different bioinformatics tools than other papers publish. The difference in algorithm might contribute to this inconsistency. Secondly, it could be due to sequence quality of different datasets. Although the sequencing protocols are the same, the tissue samples used in the experiment could have different qualities. Thirdly, the knockout methods are different in these two studies. The Blackshaw lab uses CRISPR to knockout PTBP1; the Fu lab uses viral infection to knockout PTBP1. Although the two knockout methods both succeeded, we yet to know if they have other effects in the expression profiles of the brain.\n",
    "\n",
    "Further analysis could be useful to understand if knockout PTBP1 has universal effect in other parts of the nervous system. Previous papers have shown that knockout PTBP1 could convert glial cells in the eyes into functional rod cells important in vision. This is important to treat diseases like cataract. these datasets (wild-type and PTBP1 knockout single cell sequencing) are also available publicly, and we can analyze them to understand if the conversion from glial cells to neurons also happen in other parts of the nervous system. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Team Contributions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "- Carlos Garcia - Bckground, EDA, Video Presentation, Data Cleaning, Data Setup\n",
    "- Sai Hosuru - Background, EDA, Data Cleaning, Slides for Video Presentation \n",
    "- Qilin Zhao - Background, Data Cleaning, EDA, Datasets, Data Setup, Slides for Video Presentation \n",
    "- Wilson Lubeck - Background, EDA, Data Cleaning, Slides for Video Presentation \n",
    "- Sharon Zhao - Background, EDA, Data Cleaning, Slides for Video Presentation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.6 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  },
  "vscode": {
   "interpreter": {
    "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
